code/game/objects/items/devices/gps.dm

```diff
@@ -1,31 +1,388 @@
+		gps_data["degrees"] =   round(Get_Angle(curr,T))
+		gps_data["distX"] =     T.x - curr.x
+		gps_data["distY"] =     T.y - curr.y
+		gps_data["distance"] =  round(get_dist(curr, T), 10)
+		gps_data["local"] =     (curr.z == T.z)
+		gps_data["x"] =         T.x
+		gps_data["y"] =         T.y
+		gps_list += list(gps_data)
+
+	if(length(gps_list))
+		.["gps_list"] = gps_list
+	else
+		.["no_signals"] = TRUE
+
+/obj/item/gps/ui_interact(mob/user, ui_key = "main", datum/nanoui/ui = null, force_open = 1, var/master_ui = null, var/datum/topic_state/state = global.default_topic_state)
+
+	var/data = ui_data()
+	ui = SSnano.try_update_ui(user, src, ui_key, ui, data, force_open)
+	if(!ui)
+		ui = new(user, src, ui_key, "gps.tmpl", name, 500, 300, master_ui = master_ui, state = state)
+		ui.set_initial_data(data)
+		ui.open()
+		ui.set_auto_update(1)
+
+/obj/item/gps/OnTopic(var/mob/user, var/list/href_list)
+
+	if(href_list["toggle_power"])
+		toggle_tracking()
+		. = TOPIC_REFRESH
+
+	if(href_list["track_label"])
+		var/gps_ref = href_list["track_label"]
+		var/obj/item/gps/gps = locate(gps_ref)
+		if(istype(gps) && !QDELETED(gps) && !LAZYACCESS(showing_tracked_names, gps_ref))
+			LAZYSET(showing_tracked_names, gps_ref, TRUE)
+		else
+			LAZYREMOVE(showing_tracked_names, gps_ref)
+		to_chat(user, SPAN_NOTICE("\The [src] is [LAZYACCESS(showing_tracked_names, gps_ref) ? "now showing" : "no longer showing"] labels for [gps.gps_tag]."))
+		. = TOPIC_REFRESH
+
+	if(href_list["stop_track"])
+		var/gps_ref = href_list["stop_track"]
+		var/obj/item/gps/gps = locate(gps_ref)
+		compass.clear_waypoint(gps_ref)
+		LAZYREMOVE(tracking_devices, gps_ref)
+		LAZYREMOVE(showing_tracked_names, gps_ref)
+		if(istype(gps) && !QDELETED(gps))
+			to_chat(user, SPAN_NOTICE("\The [src] is no longer tracking [gps.gps_tag]."))
+		update_compass()
+		. = TOPIC_REFRESH
+
+	if(href_list["start_track"])
+		var/gps_ref = href_list["start_track"]
+		var/obj/item/gps/gps = locate(gps_ref)
+		if(istype(gps) && !QDELETED(gps))
+			LAZYSET(tracking_devices, gps_ref, COLOR_SILVER)
+			LAZYSET(showing_tracked_names, gps_ref, TRUE)
+			to_chat(user, SPAN_NOTICE("\The [src] is now tracking [gps.gps_tag]."))
+			update_compass()
+			. = TOPIC_REFRESH
+
+	if(href_list["track_color"])
+		var/obj/item/gps/gps = locate(href_list["track_color"])
+		if(istype(gps) && !QDELETED(gps))
+			var/new_colour = input("Enter a new tracking color.", "GPS Waypoint Color") as color|null
+			if(new_colour && istype(gps) && !QDELETED(gps) && holder == user && !user.incapacitated())
+				to_chat(user, SPAN_NOTICE("You adjust the colour \the [src] is using to highlight [gps.gps_tag]."))
+				LAZYSET(tracking_devices, href_list["track_color"], new_colour)
+				update_compass()
+				. = TOPIC_REFRESH
+
+	if(href_list["tag"])
+		var/a = input("Please enter desired tag.", name, gps_tag) as text
+		a = uppertext(copytext(sanitize(a), 1, 11))
+		if(in_range(src, user))
```

Comment:
`CanInteract`, probably. If you _actually_ want what this proc is doing (doubtful) use `CanPhysicallyInteract`, I guess.